# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmSignalHandlers.cpp
# Commit: f981949ae991
# Full Hash: f981949ae991960c2099ea85435ddbb660a4229c
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-10-31 21:05:17
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 3: Implement the algorithm for obtaining tls and use it for wasm signal handling. r=lth
#   
#   This is the third part of series of patches to Frame without tls pointer.
#   Here we preserve initial tls in all entry stubs and then use it to find a proper tls instance for a given frame.
#   
# ==============================================================================

diff -r 6d85470024c3 -r f981949ae991 js/src/wasm/WasmSignalHandlers.cpp
--- a/js/src/wasm/WasmSignalHandlers.cpp	Sat Oct 31 09:56:05 2020 +0000
+++ b/js/src/wasm/WasmSignalHandlers.cpp	Sat Oct 31 09:56:05 2020 +0000
@@ -717,7 +717,9 @@
   // due to this trap occurring in the indirect call prologue, while fp points
   // to the caller's Frame which can be in a different Module. In any case,
   // though, the containing JSContext is the same.
-  Instance* instance = ((Frame*)ContextToFP(context))->instance();
+
+  auto* frame = reinterpret_cast<Frame*>(ContextToFP(context));
+  Instance* instance = GetNearestEffectiveTls(frame)->instance;
   MOZ_RELEASE_ASSERT(&instance->code() == &segment.code() ||
                      trap == Trap::IndirectCallBadSig);
 